keps/sig-storage/0033-20181111-transfer-pvc-between-namespaces.md

```diff
@@ -0,0 +1,179 @@
+
+### Goals
+
+* Provide a secure and standard mechanism to duplicate data sets on volumes across namespaces in the cluster
+* Enhance existing API's and workflows to alow transferring of volumes across namespaces
+
+### Non-Goals
+
+  While this concept could be extended, volumes are a special case.  Most other resources can just be recreated easily, but this isn't as trivial with large data sets.
+* Introducing a new API
+* Introducing a new Transfer Object
+* Enable transfer for resources other than volumes
+
+## Proposal
+
+
+Update the PVC with a transfer request annotation:
+```yaml
+apiversion: v1
+kind: PersistentVolumeClaim
+metadata:
+    name: pvc-foo
+	annotations:
+	    transferRequest: development/pvc-fromfoo
+...
+```
+
+The transferRequest update, will be detected by the PV controller, if the transfer feature is enabled, the controller will mark the PV backing the PVC for transfer.  This notation is used when the PVC is deleted from the original namespace.  The transfer request overrides the reclaim policy of the PV, and upon deletion of the claim from the original namespace, the claimref is updated to point to the new namespace/pvc-name, and the PV is not deleted.
+
+The result is annotations added to the PV and PVC by the controller:
+```yaml
+apiversion: v1
+kind: PersistentVolume
+metadata:
+    name: pvc-fromfoo
+	annotations:
+	    pv.kubernetes.io/transfer-status: pending
+		pv.kubernetes.io/transfer-source: default/pvc-foo
+		pv.kubernetes.io/transfer-destination: development/pvc-fromfoo
+...
+```
+
+```yaml
+apiversion: v1
+kind: PersistentVolumeClaim
+metadata:
+    name: pvc-fromfoo
+	annotations:
+	    transferAccept: default/pvc-foo
+	    pvc.kubernetes.io/transfer-status: pending
+...
+```
+
+The transfer-status will remain in a ``pending`` state until:
+1. The original user deletes their claim
+2. The destination user requests a claim with the proper name and annotations
+
+User in the development namespace requests a claim that matches the originators transfer request:
+```yaml
+apiversion: v1
+kind: PersistentVolumeClaim
+metadata:
+    name: pvc-fromfoo
+	annotations:
+	    transferAccept: default/pvc-foo
+...
+```
+
+The claim request in the destination namespace will be ``pending`` until the source user deletes claim ``pvc-foo`` from their namespace.  The source user can also delete the ``pvc-foo`` from their namespace before the claim request is made in the destination namespace as well.   Note that the sequence described preserves integrity of our users quotas on both sides, and even if a user in the reveiving namespace requests a claim with the names specified in a transfer request, the claim is ONLY fulfilled by that backing PV is they specified a ``transferAccept``.  This prevents accidental acceptance of transfers.
+
+Upon success, the PV annotation: ``pv.kubernetes.io/transfer-status`` annotation will be updated by the controller to ``complete``.
+
+### User Stories [optional]
```

Comment:
The user stories here do not match the title and description for the KEP. Can you please revise?

Reply:
@mattfarina Sorry, what doesn't match?  The title and purpose of the KEP is to enable transferring a PVC from one namespace to another.  The user stories below are specific concrete cases.  When cloning we create a new PVC, snapshots are restored by creating a new PVC from a Snapshot and the last case perhaps I have a namespace with users that have special tools for populating/generating data that I then want to "give" to another namespace/user to consume.

Let me know how I can make this better, I'm really not sure where the disconnect is currently.

Reply:
Upon  re-reading, I see where you are going with this.

The debug comment you had, elsewhere, does a good job explaining the user story using a concrete example. Could you put that under the appropriate user story to add context.